code/datums/stat_logger/logger.dm

```diff
@@ -0,0 +1,147 @@
+/* ._.-'~'-._.-'~'-._.-'~'-._.-'~'-._.-'~'-._. */
+/*-=-=-=-=-=-=-=STAT LOGGER BY WIRE-=-=-=-=-=-=*/
+/* '~'-._.-'~'-._.-'~'-._.-'~'-._.-'~'-._.-'~' */
+
+/// Holder datum for [/datum/stats]
+/var/global/datum/stats/stats
+
+/**
+ * # Stats
+ *
+ * Holder datum for statistic event logging
+ *
+ * Handles building stat events with metadata, and formatting them in such a
+ * way that the backend stat parsing service can ingest
+ *
+ * Usage:
+ * ```
+ * stats.record("bee_farts", list("bee_name" = "Tootington the Third"))
+ * ```
+ */
+/datum/stats
+	/// The base path for all stat log files
+	var/stats_path = "data/stats"
+	/// The file handler for stat logging this round
+	var/log_file
+	/// A queue of stat records we couldn't save immediately (probably because we didn't have a round ID yet)
+	var/list/queue = list()
+	/// A list of foreign relationship definitions
+	var/list/foreign_relations = list()
+
+	New()
+		..()
+		// Build out all the foreign relationship definitions ahead of time
+		for (var/type in typesof(/datum/stats_foreign))
+			if (type == /datum/stats_foreign) continue
```

Comment:
use `concrete_typesof(/datum/stats_foreign)`
and `ABSTRACT_TYPE(/datum/stats_foreign)`